### PR TITLE
[`Refactor`] `StableDiffusionReferencePipeline` inheriting from `DiffusionPipeline`

### LINKED ISSUE #6984: refactors pipelines inheriting from `StableDiffusionPipeline`
We have a couple of community pipelines that used `StableDiffusionPipeline` as the base class instead of `DiffusionPipeline`. Most of them are broken after we changed the signature of `StableDiffusionPipeline`. see more details on the issue here https://github.com/huggingface/diffusers/issues/6969#issuecomment-1944162819

pipelines should not use `StableDiffusionPipeline` as a base class. Is anyone interested in refactoring some of these community pipelines to inherit from `DiffusionPipeline` instead? You will need to:
1. update the base class to use `DiffusionPipeline` and any of the mixins needed e.g. 
```python
class CommunityPipeline(
    DiffusionPipeline, TextualInversionLoaderMixin, LoraLoaderMixin, IPAdapterMixin, FromSingleFileMixin
):
```
2. if the community pipeline use some of the methods in `StableDiffusionPipeline` - you need use `#Copied from` statement to copy over methods to your community pipeline. Learn more about the `Copied from` mechanism here https://huggingface.co/docs/diffusers/conceptual/contribution#copied-from-mechanism

Here is the list of pipelines that need to be refactored!
- [ ] Prompt2PromptPipeline(@ihkap11 is working on it): https://github.com/huggingface/diffusers/blob/main/examples/community/pipeline_prompt2prompt.py
- [ ] RegionalPromptingStableDiffusionPipeline (@pranjalks is working on it): https://github.com/huggingface/diffusers/blob/main/examples/community/regional_prompting_stable_diffusion.py
- [x] StableDiffusionReferencePipeline ( completed by @standardAI https://github.com/huggingface/diffusers/pull/7071)
- [ ] TensorRTStableDiffusionPipeline (@Bhavay-2001 is working on it): https://github.com/huggingface/diffusers/blob/main/examples/community/stable_diffusion_tensorrt_txt2img.py

### PR DESCRIPTION
**Pull request by standardAI**
Fixes #6984.

This pull request refactors the `StableDiffusionReferencePipeline` class to inherit from the `DiffusionPipeline` class rather than the `StableDiffusionPipeline` class.
Again, thank you for this opportunity. I tried with [the code example](https://github.com/huggingface/diffusers/blob/main/examples/community/README.md#stable-diffusion-reference), and it seems to work. I would be glad to have your advice/feedback.
@yiyixuxu

### CODE REVIEW

**Comment by yiyixuxu** [comment]
@okotaku 
can you also take a look here? 
@standardAI refactored the pipeline so it inherits from `DiffusionPipeline`

**Review by yiyixuxu — APPROVED** [review]
thank you!
looks very nice to me! thanks for adding these doc string too

**Comment by okotaku** [comment]
@standardAI Good job👍
LGTM!

**Comment by standardAI** [comment]
Thanks for reviewing and merging!

**Comment by Bhavay-2001** [comment]
Hi @standardAI, did you tried running the code example on colab notebooks or local machine?

**Comment by standardAI** [comment]
> Hi @standardAI, did you tried running the code example on colab notebooks or local machine?

Hi @Bhavay-2001. I used Colab with `torch_dtype=torch.float16` and `variant='fp16'` parameters.

**Comment by Bhavay-2001** [comment]
Hi @standardAI, could you please share your notebook here. 
Thanks

**Comment by standardAI** [comment]
> Hi @standardAI, could you please share your notebook here. Thanks

[Here.](https://colab.research.google.com/gist/standardAI/97a84796473aaf9905877aaaf4c2bc9c/stablediffusionreferencepipeline.ipynb)

**Comment by Bhavay-2001** [comment]
Hi @standardAI, could you please review my notebook [here](https://colab.research.google.com/drive/1Z4tjt3J7xKRb6eMmzJMcLXN77_QQjXFT?usp=sharing). I am running it on colab only but it seems to run out of memory every time. Could you please suggest what am I missing?

**Comment by standardAI** [comment]
> Hi @standardAI, could you please review my notebook [here](https://colab.research.google.com/drive/1Z4tjt3J7xKRb6eMmzJMcLXN77_QQjXFT?usp=sharing). I am running it on colab only but it seems to run out of memory every time. Could you please suggest what am I missing?

As a first look, did you choose T4 GPU in the Runtime tab?

**Comment by Bhavay-2001** [comment]
Yes I did.

**Comment by standardAI** [comment]
> Yes I did. The current state of the notebook is when I tried with CPU power because I was crossed my computing limit on colab

Hmm, OK then. I will examine it in detail within a few days.

File: examples/community/stable_diffusion_reference.py [hunk 1]
```diff
@@ -158,6 +775,35 @@ def prepare_ref_latents(self, refimage, batch_size, dtype, device, generator, do
         ref_image_latents = ref_image_latents.to(device=device, dtype=dtype)
         return ref_image_latents
 
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.run_safety_checker
```
Comment (standardAI) [inline examples/community/stable_diffusion_reference.py:778]: @Bhavay-2001 for example.

**Pull request merged**